src/cdk/testing/harness-environment.ts

```diff
@@ -147,43 +130,179 @@ export abstract class HarnessEnvironment<E> implements HarnessLoader, LocatorFac
    */
   protected abstract getAllRawElements(selector: string): Promise<E[]>;
 
-  private async _getAllHarnesses<T extends ComponentHarness>(
-      harnessType: ComponentHarnessConstructor<T> | HarnessPredicate<T>): Promise<T[]> {
-    const harnessPredicate = harnessType instanceof HarnessPredicate ?
-        harnessType : new HarnessPredicate(harnessType, {});
-    const elements = await this.getAllRawElements(harnessPredicate.getSelector());
-    return harnessPredicate.filter(elements.map(
-        element => this.createComponentHarness(harnessPredicate.harnessType, element)));
+  /** Gets a list of matching harnesses and test elements for a given query list. */
+  private async _getAllHarnessesAndTestElements<T extends (HarnessQuery<any> | string)[]>(
+      queries: T): Promise<LocatorFnResult<T>[]> {
+    const {allQueries, harnessQueries, elementQueries, harnessTypes} = _parseQueries(queries);
+    // Combine all of the queries into one large comma-delimited selector and use it to get all raw
+    // elements matching any of the individual queries.
+    const rawElements = await this.getAllRawElements(
+        [...elementQueries, ...harnessQueries.map(pred => pred.getSelector())].join(','));
```

Comment:
If you combine everything into one big selector, will that still follow the order the queries were provided?

Reply:
We want to sort first by DOM element order and then secondarily by query order. Since this will only find max 1 result per DOM element, the query order isn't important here. Later when we match DOM elements to harnesses we'll worry about the query order.